File: cpp-answers/67.二进制求和.cpp
/*
 * @Author: your name
 * @Date: 2020-05-30 01:04:34
 * @LastEditTime: 2020-05-30 01:07:59
 * @LastEditors: your name
 * @Description: In User Settings Edit
 * @FilePath: \leetcode刷题\67.二进制求和.cpp
 */
/*
 * @lc app=leetcode.cn id=67 lang=cpp
 *
 * [67] 二进制求和
 */

// @lc code=start
class Solution
{
public:
    string addBinary(string a, string b)
    {
        int i = a.size() - 1, j = b.size() - 1;
        int carry = 0;
        string ret = "";
        while (i >= 0 || j >= 0 || carry != 0)
        {
            if (i >= 0)
            {
                carry += a[i] == '0' ? 0 : 1;
                i--;
            }
            if (j >= 0)
            {
                carry += b[j] == '0' ? 0 : 1;
                j--;
            }
            ret = ((carry % 2) == 0 ? "0" : "1") + ret;
            carry /= 2;
        }
        return ret;
    }
};
// @lc code=end

File: rust-answers/67.二进制求和.rs
/*
 * @lc app=leetcode.cn id=67 lang=rust
 *
 * [67] 二进制求和
 */

// @lc code=start
impl Solution {
    pub fn add_binary(a: String, b: String) -> String {
        // format!("{:b}", u128::from_str_radix(&a, 2).unwrap() + u128::from_str_radix(&b, 2).unwrap())
        let ia = u128::from_str_radix(&a, 2).unwrap();
        let ib = u128::from_str_radix(&b, 2).unwrap();
        format!("{:b}", ia + ib)
    }
}
// @lc code=end

